service.py

```diff
@@ -20,56 +23,37 @@ def handler(event, context):
 
     if DATASERVICE is None:
         return 'no dataservice url set'
-    updater = AclUpdater(DATASERVICE)
+    updater = AclUpdater(DATASERVICE, context)
     res = {}
-    failed = {'Records': []}
-    for i, record in enumerate(event['Records']):
-
-        # If we're running out of time, stop processing and re-invoke
-        # NB: We check that i > 0 to ensure that *some* progress has been made
-        # to avoid infinite call chains.
+    while len(event['Records']) > 0:
         if (hasattr(context, 'invoked_function_arn') and
-            context.get_remaining_time_in_millis() < 150000 and
-                i > 0):
-            records = event['Records'][i:]
+                context.get_remaining_time_in_millis() < 15000):
             print('not able to complete {} records, '
-                  're-invoking the function'.format(len(records)))
-            remaining = {'Records': records}
+                  're-invoking the function'.format(len(event['Records'])))
             lam = boto3.client('lambda')
             # Invoke the lambda again with remaining records
             response = lam.invoke(
                 FunctionName=context.invoked_function_arn,
                 InvocationType='Event',
-                Payload=str.encode(json.dumps(remaining))
+                Payload=str.encode(json.dumps(event))
             )
             # Stop processing and exit
             break
         else:
-            # update consent code and acl
-            failed_record = updater.update_acl(record)
-            if not failed_record:
-                failed['Records'].append(record)
-            else:
-                res['genomic_file'] = 'processed all records'
-
-    if failed['Records'] and hasattr(context, 'invoked_function_arn'):
-        lam = boto3.client('lambda')
-        # Invoke the lambda again with failed records
-        response = lam.invoke(
-            FunctionName=context.invoked_function_arn,
-            InvocationType='Event',
-            Payload=str.encode(json.dumps(failed))
-        )
-        res['genomic_file'] = 'calling lambda for failed records'
-    else:
-        res['genomic_file'] = 'processed all records'
+            try:
+                record = event['Records'].pop()
+                updater.update_acl(record)
+                res["genomic_file"] = 'processed all records'
+            except TimeoutException:
```

Comment:
I think you should catch any exception, not just TimeoutException. requests will raise ValueError if the timeout is set to 0 or less, which can happen if self.context.get_remaining_time_in_millis returns the same or less than the subtracted buffer, which is basically a good sign that you should restart the lambda right away.